lib/evaluate/intrinsics.cc

```diff
@@ -1390,33 +1391,55 @@ std::optional<SpecificCall> IntrinsicProcTable::Implementation::Probe(
     return result;
   }
   // Probe the specific intrinsic function table first.
-  parser::Messages specificBuffer;
-  parser::ContextualMessages specificErrors{
-      call.name, finalBuffer ? &specificBuffer : nullptr};
-  FoldingContext specificContext{context, specificErrors};
+  parser::Messages localBuffer, specificBuffer;
+  parser::ContextualMessages localMessages{
+      call.name, finalBuffer ? &localBuffer : nullptr};
+  FoldingContext localContext{context, localMessages};
   std::string name{call.name.ToString()};
   auto specificRange{specificFuncs_.equal_range(name)};
   for (auto iter{specificRange.first}; iter != specificRange.second; ++iter) {
     if (auto specificCall{
-            iter->second->Match(call, defaults_, arguments, specificContext)}) {
+            iter->second->Match(call, defaults_, arguments, localContext)}) {
       if (const char *genericName{iter->second->generic}) {
         specificCall->specificIntrinsic.name = genericName;
       }
       specificCall->specificIntrinsic.isRestrictedSpecific =
           iter->second->isRestrictedSpecific;
+      if (finalBuffer != nullptr) {
+        finalBuffer->Annex(std::move(localBuffer));
+      }
       return specificCall;
+    } else {
+      specificBuffer.Annex(std::move(localBuffer));
     }
   }
   // Probe the generic intrinsic function table next.
   parser::Messages genericBuffer;
-  parser::ContextualMessages genericErrors{
-      call.name, finalBuffer ? &genericBuffer : nullptr};
-  FoldingContext genericContext{context, genericErrors};
   auto genericRange{genericFuncs_.equal_range(name)};
   for (auto iter{genericRange.first}; iter != genericRange.second; ++iter) {
     if (auto specificCall{
-            iter->second->Match(call, defaults_, arguments, genericContext)}) {
+            iter->second->Match(call, defaults_, arguments, localContext)}) {
+      // Apply any semantic checks peculiar to the intrinsic
+      if (call.name == "present") {
+        bool ok{false};
+        if (const auto &arg{specificCall->arguments[0]}) {
+          if (const auto *expr{arg->GetExpr()}) {
+            if (const Symbol * symbol{GetLastSymbol(*expr)}) {
```

Comment:
From what I understand, this means that the following will be allowed:
```
subroutine(a)
  type :: t
    real, allocatable :: x(:)
  end type
  type(t), optional :: a(*)
  
  if (PRESENT(a(2)%x)) then  ! Equivalent to PRESENT(a)
    ! ...
  end if
end subroutine
```
Is this the intention to allow such usage ? I am not sure where it could lead to issues in user program, but maybe there is a risk of bad IR generation in inlined PRESENT implementation (though it can be made to handle this).

Reply:
Ah, yes, but `x` won't have the `OPTIONAL` attribute.

But I need to tighten that test up to disallow `PRESENT(a(2))`.